rust/lance-encoding/src/data.rs

```diff
@@ -290,6 +290,48 @@ impl FixedWidthDataBlock {
     }
 }
 
+pub struct VariableWidthDataBlockBuilder1 {
+    offsets: Vec<u32>,
+    bytes: Vec<u8>,
+}
+
+impl VariableWidthDataBlockBuilder1 {
+    fn new(estimated_size_bytes: u64) -> Self {
+        Self {
+            offsets: vec![0u32],
+            bytes: Vec::with_capacity(estimated_size_bytes as usize),
+        }
+    }
+}
+
+impl DataBlockBuilderImpl for VariableWidthDataBlockBuilder1 {
+    fn append(&mut self, data_block: &mut DataBlock, selection: Range<u64>) {
+        let block = data_block.as_variable_width_mut_ref().unwrap();
+        assert!(block.bits_per_offset == 32);
+
+        let offsets = block.offsets.borrow_to_typed_slice::<u32>();
+        let offsets = offsets.as_ref();
+
+        let start_offset = offsets[selection.start as usize];
+        let end_offset = offsets[selection.end as usize];
+        self.bytes
+            .extend_from_slice(&block.data[start_offset as usize..end_offset as usize]);
+        self.offsets.push(self.bytes.len() as u32);
```

Comment:
I'm not sure this is correct.  Shouldn't you be extending from `block.offsets` and not pushing a single value?